pkg/streamingpromql/operators/functions/absent.go

```diff
@@ -0,0 +1,144 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/functions.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors.
+
+package functions
+
+import (
+	"context"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/promql/parser/posrange"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// Absent is an operator that implements the absent() function.
+type Absent struct {
+	timeRange                types.QueryTimeRange
+	argExpressions           parser.Expr
+	inner                    types.InstantVectorOperator
+	expressionPosition       posrange.PositionRange
+	memoryConsumptionTracker *limiting.MemoryConsumptionTracker
+	presence                 []bool
+}
+
+var _ types.InstantVectorOperator = &Absent{}
+
+// NewAbsent creates a new Absent.
+func NewAbsent(inner types.InstantVectorOperator, innerExpr parser.Expr, timeRange types.QueryTimeRange, expressionPosition posrange.PositionRange, memoryConsumptionTracker *limiting.MemoryConsumptionTracker) *Absent {
+	return &Absent{
+		timeRange:                timeRange,
+		inner:                    inner,
+		argExpressions:           innerExpr,
+		expressionPosition:       expressionPosition,
+		memoryConsumptionTracker: memoryConsumptionTracker,
+	}
+}
+
+func (a *Absent) SeriesMetadata(ctx context.Context) ([]types.SeriesMetadata, error) {
+	innerMetadata, err := a.inner.SeriesMetadata(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer types.PutSeriesMetadataSlice(innerMetadata)
+
+	a.presence, err = types.BoolSlicePool.Get(a.timeRange.StepCount, a.memoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	// Initialize presence slice
+	a.presence = a.presence[:a.timeRange.StepCount]
+
+	metadata := types.GetSeriesMetadataSlice(1)
+	metadata = append(metadata, types.SeriesMetadata{
+		Labels: createLabelsForAbsentFunction(a.argExpressions),
+	})
+
+	for range innerMetadata {
+		series, err := a.inner.NextSeries(ctx)
+		if err != nil {
+			return metadata, err
```

Comment:
For safety, if we're returning an error, we shouldn't return the metadata slice:

```suggestion
			return nil, err
```